### PR TITLE
namesys: add entry to DHT cache after publish

### LINKED ISSUE #2078: name publish doesn't instantly update the cache entry in the local node
When `ipfs name publish` finishes, it doesn't instantly update the cache of the local node.

This is a problem in [my app](https://github.com/fazo96/ipfs-boards): the user updates his profile, but it takes a while (around 20 seconds) until he sees his changes.

### PR DESCRIPTION
**Pull request by Kubuxu**
Resolves #2078

License: MIT
Signed-off-by: Jakub Sztandera <kubuxu@protonmail.ch>

### CODE REVIEW

**Comment by Kubuxu** [comment]
@whyrusleeping it isn't supper nice solution but without some refactor there isn't anything much better might be done.

File: namesys/namesys.go [hunk 1]
```diff
@@ -87,9 +89,44 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
 
 // Publish implements Publisher
 func (ns *mpns) Publish(ctx context.Context, name ci.PrivKey, value path.Path) error {
-	return ns.publishers["/ipns/"].Publish(ctx, name, value)
+	err := ns.publishers["/ipns/"].Publish(ctx, name, value)
+	if err != nil {
+		return err
+	}
+	ns.addToDHTCache(name, value, time.Now().Add(time.Hour*24))
```
Comment (whyrusleeping) [inline namesys/namesys.go]: lets pull this magic number out and justify it

File: namesys/namesys.go [hunk 2]
```diff
@@ -87,9 +89,44 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
...
+	return nil
 }
 
-func (ns *mpns) PublishWithEOL(ctx context.Context, name ci.PrivKey, val path.Path, eol time.Time) error {
-	return ns.publishers["/ipns/"].PublishWithEOL(ctx, name, val, eol)
+func (ns *mpns) addToDHTCache(key ci.PrivKey, value path.Path, eol time.Time) {
+	var err error
+	value, err = path.ParsePath(value.String())
+	if err != nil {
+		log.Error("could not parse path")
```
Comment (whyrusleeping) [inline namesys/namesys.go]: maybe add the actual error to this? might aid in debugging random error messages

File: namesys/namesys.go [hunk 3]
```diff
@@ -87,9 +89,44 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
...
+	}
+	ns.addToDHTCache(name, value, eol)
+	return nil
 }
 
-func (ns *mpns) PublishWithEOL(ctx context.Context, name ci.PrivKey, val path.Path, eol time.Time) error {
-	return ns.publishers["/ipns/"].PublishWithEOL(ctx, name, val, eol)
+func (ns *mpns) addToDHTCache(key ci.PrivKey, value path.Path, eol time.Time) {
+	var err error
+	value, err = path.ParsePath(value.String())
```
Comment (whyrusleeping) [inline namesys/namesys.go]: wait, why are we converting the thing to a string, then back to a path?

File: namesys/namesys.go (hunk 1 above)
Comment (Kubuxu) [inline namesys/namesys.go]: It is the same as in publisher.go, should I add comment for it?

File: namesys/namesys.go (hunk 3 above)
Comment (Kubuxu) [inline namesys/namesys.go]: Yes as the publisher can publish hash only but resolver will sanitize this value to be `/ipfs/CID`.

File: namesys/namesys.go (hunk 2 above)
Comment (Kubuxu) [inline namesys/namesys.go]: This is only cache, I didn't want it to fail the publish (as it was already published and we can't revert it).

File: namesys/namesys.go (hunk 1 above)
Comment (Kubuxu) [inline namesys/namesys.go]: Or extract it from publisher?

**Comment by Kubuxu** [comment]
Yes, they are separate modules that don't know about echo other's existence.

**Comment by wigy-opensource-developer** [comment]
Is this problem IPNS specific? I love how this PR solves the problem in hand without changing much code, but I see it as a temporary fix that will need to be removed after the real issue has been addressed.

The caching could be moved into the DHT-based IPRS implementation, so the /pk/ namespace could benefit from it as well. Namesys uses both. I am not aware of other IPRS namespaces, but probably it would apply to any record types.

**Comment by Kubuxu** [comment]
IPRS doesn't exist yet, I like to call IPNS a huge PoC that mutable data is possible even though the content is immutable.

Yes, after IPRS is designed, speced out, and implemented it can be moved into one place with infrastructure that will include caching of self published record in its design. When code of namesys was written it wasn't a feature that was taken into account.

File: namesys/namesys.go (hunk 1 above)
Comment (whyrusleeping) [inline namesys/namesys.go]: I'd extract it from both, and give it a small comment

Comment (whyrusleeping) [inline namesys/namesys.go]: Lets just call it `DefaultRecordTTL` (i had to go look at it again to remember what exactly this was)

Comment (Kubuxu) [inline namesys/namesys.go]: Hah, gave it the same name without reading this. <3

File: namesys/namesys.go [hunk 4]
```diff
@@ -87,9 +89,47 @@ func (ns *mpns) resolveOnce(ctx context.Context, name string) (path.Path, error)
 
 // Publish implements Publisher
 func (ns *mpns) Publish(ctx context.Context, name ci.PrivKey, value path.Path) error {
-	return ns.publishers["/ipns/"].Publish(ctx, name, value)
+	err := ns.publishers["/ipns/"].Publish(ctx, name, value)
+	if err != nil {
+		return err
+	}
+	ns.addToDHTCache(name, value, time.Now().Add(DefaultRecortTTL))
```
Comment (whyrusleeping) [inline namesys/namesys.go]: typo: DefaultRecor<b>d</b>TTL

**Comment by Cole128** [comment]
I am still experiencing slow ipns resolving. For a minute after I publish it will be fast, but then it takes a very long time.

**Pull request merged**